### DISCUSSION TITLE
Unable to compile CUDA codes after macos-10.15 retirement

### DISCUSSION THREAD

**Discussion by fangq** [actions/virtual-environments#7838] [General questions] [answered]
we have previously set up an action script to build a CUDA based software on the `macos-10.15` runner

https://github.com/fangq/mcx/blob/master/.github/workflows/build_macos_wheel.yml
https://github.com/fangq/mcx/blob/master/.github/workflows/build_all.yml

however, since github retires `macos-10.15`, our script longer works. I've spent a significant time looking for ways to compile cuda on `macos-11` or newer, but non of those works. nvcc, regardless what version I use, always complain that Apple Clang or GNU GCC is not supported on the system.

since NVIDIA ended CUDA support on mac since CUDA 10.2, see

https://www.provideocoalition.com/officially-official-nvidia-drops-cuda-support-for-macos/

I've tried older versions of CUDA, but still encounter the same error

please advise how to build CUDA projects on the macos runners, thanks

**Comment by vpolikarpov-akvelon** [comment]
Hey @fangq. It appears that the latest version of the CUDA Toolkit compatible with macOS is indeed 10.2. This version is only compatible with macOS up to version 10.13 and is not suitable for macOS 11 and 10.14+.

Since new versions of macOS are not compatible with the CUDA Toolkit, running the workload on your own machine that operates on macOS High Sierra would be the most feasible solution. By utilizing a self-hosted machine, you should be able to continue your work without any compatibility issues.

**Comment by fangq** [comment]
@vpolikarpov-akvelon, thanks for your reply. 

I understood that both Apple and NVIDIA had abandoned each other - so that future Apple silicon no longer uses NVIDIA or AMD GPUs, even discrete PCIe cards, and NVIDIA stopped shipping drivers to Mac OS, however, these decisions were made fairly recently, and there is still a large body of intel Mac hardware existing, with PCIe or eGPU capacity to run NVIDIA GPUs.

Based on our tracking, Mac users still consist a significant (20-30%) portion of my CUDA software downloads. Because of this, we really hope to continue testing and building Mac binaries for this user group. Currently we use github action to build and update a binary Python module for Linux/Windows/Mac. However, since the retirement of macos-10.13, our mac modules could no longer be built.

Comparing the runner versions provided by github and other competitive CI providers (such as Travis [Linux](https://docs.travis-ci.com/user/reference/linux/) and [Mac](https://docs.travis-ci.com/user/reference/osx/) runners), the supported version is quite limited on github. what are the main concerns for supporting mainstream/post-mainstream OS versions? I felt that github's runner versions are overall too new to build and test software for the bulk of the computing hardware that is currently available.

**Reply by vpolikarpov-akvelon** [reply, marked as answer]
Well, there is nothing we can do, unfortunately, and I'm really sorry for that. The image `macos-10.13` was deprecated as per policy described here: https://github.com/actions/runner-images#image-deprecation-policy. It's only possible now to run GitHub CI workflows on mac OS 10.13 using self-hosted runners.
